src/instruments.rs

```diff
@@ -1,133 +1,331 @@
+    fn all_templates(&self) -> Vec<&str> {
+        self.standard_templates
+            .iter()
+            .chain(self.custom_templates.iter())
+            .map(|value| value.as_str())
+            .collect::<Vec<_>>()
+    }
+}
+
+/// Represents the Xcode Instrument version detected.
+pub enum XcodeInstruments {
+    XcTrace,
+    InstrumentsBinary,
+}
+
+impl XcodeInstruments {
+    /// Detects which version of Xcode Instruments is installed and if it can be launched.
+    pub(crate) fn detect() -> Result<XcodeInstruments> {
+        let cur_version = get_macos_version()?;
+        let macos_xctrace_version = Version::parse("10.15.0").unwrap();
+
+        if cur_version >= macos_xctrace_version {
+            // This is the check used by Homebrew,see
+            // https://github.com/Homebrew/install/blob/master/install.sh#L216
+            let clt_git_filepath = Path::new("/Library/Developer/CommandLineTools/usr/bin/git");
+            if clt_git_filepath.exists() {
+                return Ok(XcodeInstruments::XcTrace);
+            }
+        } else {
+            let instruments_app_filepath = Path::new("/usr/bin/instruments");
+            if instruments_app_filepath.exists() {
+                return Ok(XcodeInstruments::InstrumentsBinary);
+            }
+        }
         Err(anyhow!(
-            "/usr/bin/instruments does not exist. \
-             Please install the Xcode Command Line Tools."
+            "Xcode Instruments is not installed. Please install the Xcode Command Line Tools."
         ))
     }
+
+    /// Return a string listing available Instruments Templates.
+    ///
+    /// The typical output of the underlying command is:
+    ///
+    /// ```
+    /// == Standard Templates ==
+    /// Activity Monitor
+    /// Allocations
+    /// Animation Hitches
+    /// App Launch
+    /// Core Data
+    /// Counters
+    /// Energy Log
+    /// File Activity
+    /// Game Performance
+    /// Leaks
+    /// Logging
+    /// Metal System Trace
+    /// Network
+    /// SceneKit
+    /// SwiftUI
+    /// System Trace
+    /// Time Profiler
+    /// Zombies
+    ///
+    /// == Custom Templates ==
+    /// MyTemplate
+    /// ```
+    ///
+    /// The custom templates only appears if you have custom templates.
+    ///
+    pub(crate) fn available_templates(&self) -> Result<TemplateCatalog> {
+        let Output { status, stdout, .. } = match self {
+            XcodeInstruments::XcTrace => {
+                Command::new("xcrun").args(&["xctrace", "list", "templates"]).output()?
+            }
+            XcodeInstruments::InstrumentsBinary => {
+                Command::new("instruments").args(&["-s", "templates"]).output()?
+            }
+        };
+
+        if !status.success() {
+            return Err(anyhow!(
+                "Could not list templates. Please check your Xcode Instruments installation."
+            ));
+        }
+
+        let templates_str = String::from_utf8(stdout)?;
```

Comment:
another silly thing but I would do `std::str::from_utf8` and avoid the allocation, although it is going to be meaningless in this particular case. Good muscle memory, though?

Reply:
Thanks! I completely missed it.

Reply:
Just for completeness, `String::from_utf8` moves in the `stdout` vector of `u8`. According to the [docs](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8) this function will take care of not copying the vector. That's pretty cool! If you don't mind, I'll leave it as is.